File: rk4.py
import numpy as np

def solve_rk4(f, t0, t_end, y0, h):
    # This method solves an ODE using the Runge-Kutta 4th order method (RK4)
    # f is the differential equation
    # t0 is the start time
    # t_end is the final time
    # y0 is the initial condition at t0
    # h is the step size

    # Check for valid step size
    if h <= 0:
        raise ValueError("Step size h must be greater than zero.")

    # Number of time steps
    N = int((t_end - t0) / h)
    
    # Create time values
    t_values = np.linspace(t0, t_end, N+1)
    
    # Initialize solution array
    y_values = np.zeros_like(t_values)
    y_values[0] = y0
    
    # Iterating over each time step
    for i in range(N):
        t_n = t_values[i]
        y_n = y_values[i]

        # Compute RK4 slopes
        k1 = f(t_n, y_n)
        k2 = f(t_n + h/2, y_n + (h/2) * k1)
        k3 = f(t_n + h/2, y_n + (h/2) * k2)
        k4 = f(t_n + h, y_n + h * k3)

        # Update equation for RK4
        y_values[i+1] = y_n + (h/6) * (k1 + 2*k2 + 2*k3 + k4)
    
    return t_values, y_values

File: DESolver/RK4/rk4.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Define the differential equation
double f(double t, double i, double V0, double R, double L) {
    return (V0 - R * i) / L;
}

// RK4 function solver
void rk4(double (*f)(double, double, double, double, double), double t0, double t_end, double i0, double h, double V0, double R, double L, double *t_val, double *i_val) {
    if (h <= 0) {
        printf("Step size must be positive.\n");
        return;
    }

    int N = (int)((t_end - t0) / h); // Number of time steps

    // Initialize first values
    i_val[0] = i0;
    t_val[0] = t0;

    // Compute points in loop
    for (int i = 0; i < N; i++) {
        double t_n = t_val[i];
        double i_n = i_val[i];

        double k1 = h * f(t_n, i_n, V0, R, L);
        double k2 = h * f(t_n + h / 2, i_n + k1 / 2, V0, R, L);
        double k3 = h * f(t_n + h / 2, i_n + k2 / 2, V0, R, L);
        double k4 = h * f(t_n + h, i_n + k3, V0, R, L);

        i_val[i + 1] = i_n + (k1 + 2 * k2 + 2 * k3 + k4) / 6;
        t_val[i + 1] = t_n + h;
    }
}

